<a href="https://colab.research.google.com/github/ryanczhang7/spotifyproject/blob/master/Spotify_Data_Acquisition.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Getting HipHop Song Data from Spotify Web API using Spotipy

**Note: hiphop data gathered on 3/1 and pop data gathered on 3/6.**

In [0]:
pip install spotipy

In [0]:
import pandas as pd
import numpy as np
from pandas.io.json import json_normalize
import requests

**I thank RareLoot at Medium for helping me with some starter code for this process with his article here: https://medium.com/@RareLoot/extracting-spotify-data-on-your-favourite-artist-via-python-d58bc92a4330.**

In [0]:
# The spotipy starter code for accessing the API
import spotipy

client_id = 
client_secret = 
# To access authorised Spotify data
from spotipy.oauth2 import SpotifyClientCredentials

client_credentials_manager = SpotifyClientCredentials(
    client_id=client_id, client_secret=client_secret)
# Spotify object to access API
sp = spotipy.Spotify(client_credentials_manager=client_credentials_manager) 

In [0]:
# Searching for an artist to get his uri, which I need to get his songs
name = "Kendrick Lamar"
result = sp.search(name)
result['tracks']['items'][0]['artists']

[{'external_urls': {'spotify': 'https://open.spotify.com/artist/2YZyLoL8N0Wb9xBt1NhZWg'},
  'href': 'https://api.spotify.com/v1/artists/2YZyLoL8N0Wb9xBt1NhZWg',
  'id': '2YZyLoL8N0Wb9xBt1NhZWg',
  'name': 'Kendrick Lamar',
  'type': 'artist',
  'uri': 'spotify:artist:2YZyLoL8N0Wb9xBt1NhZWg'}]

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][0]['uri']
# Pull all of the artist's albums
sp_albums = sp.artist_albums(artist_uri, album_type='album')
# Store artist's albums' names' and uris in separate lists
album_names = []
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])
    
album_uris

['spotify:album:3pLdWdkj83EYfDN6H2N8MR',
 'spotify:album:4Kp52vGJe5oVi0EXNhSDKy',
 'spotify:album:5sOSzueqgCiVpXNcpd6QpL',
 'spotify:album:4alcGHjstaALJHHiljfy3H',
 'spotify:album:0bDN3eDKOfaeM5vccGu6SI',
 'spotify:album:1LLOGSa5ehwSWrPg1Qc3Vg',
 'spotify:album:4eLPsYPBmXABThSJ821sqY',
 'spotify:album:2ZsOtrjbcxMhcOcyDzNt1V',
 'spotify:album:0bLXUfNT34mna9aXq8ex68',
 'spotify:album:0kL3TYRsSXnu0iJvFO3rud',
 'spotify:album:7ycBtnsMtyVbbwTfJwRjSP',
 'spotify:album:76cek8jEdNsxH60aUu8Qtj',
 'spotify:album:6w7lqIsvDPgTChMrPw5oIL',
 'spotify:album:0CXmtbhJapIH7vw8go2DEb',
 'spotify:album:2dnvkiEsLtEMOAmM0DaZVo',
 'spotify:album:2ioBMJIiOANfeCaXP9a7Zn',
 'spotify:album:21EBAOQBAPvE7bMsz6sGV7',
 'spotify:album:748dZDqSZy6aPXKcI9H80u',
 'spotify:album:1DqhWr73Fh5yoNzKLas0G3',
 'spotify:album:1gItED20w7gVObqo6sSh8y']

In [0]:
album_names

['Black Panther The Album Music From And Inspired By',
 'Black Panther The Album Music From And Inspired By',
 'Black Panther The Album Music From And Inspired By',
 'DAMN. COLLECTORS EDITION.',
 'DAMN. COLLECTORS EDITION.',
 'DAMN. COLLECTORS EDITION.',
 'DAMN.',
 'DAMN.',
 'DAMN.',
 'untitled unmastered.',
 'To Pimp A Butterfly',
 'To Pimp A Butterfly',
 'To Pimp A Butterfly',
 'To Pimp A Butterfly',
 'good kid, m.A.A.d city (Deluxe)',
 'good kid, m.A.A.d city',
 'good kid, m.A.A.d city',
 'good kid, m.A.A.d city (Deluxe)',
 'good kid, m.A.A.d city (Deluxe)',
 'good kid, m.A.A.d city (Deluxe)']

In [0]:
for album in album_uris:
  print (len(sp.album_tracks(album)["items"]))

14
14
14
14
14
14
14
14
14
8
16
16
16
16
16
13
13
17
16
15


**There are many duplicates in the album_names, and when I pull up how many tracks there are in each, I see that they are duplicates. Only good kid, mAAd city has differences, so I will pick the one with the most songs. They ablums I want to get are at indices 0,3,9,10,17.**

In [0]:
# Removing duplicate albums from both lists at the same time
indices = [0,3,9,10,17]
new_album_uris = [album_uris[x] for x in indices]
new_album_names = [album_names[x] for x in indices]
new_album_names

['Black Panther The Album Music From And Inspired By',
 'DAMN. COLLECTORS EDITION.',
 'untitled unmastered.',
 'To Pimp A Butterfly',
 'good kid, m.A.A.d city (Deluxe)']

In [0]:
# Testing getting all song in an album
for song in sp.album_tracks(new_album_uris[0])["items"]:
  print(song["name"])

Black Panther
All The Stars (with SZA)
X (with 2 Chainz & Saudi)
The Ways (with Swae Lee)
Opps (with Yugen Blakrok)
I Am
Paramedic!
Bloody Waters (with Anderson .Paak & James Blake)
King's Dead (with Kendrick Lamar, Future & James Blake)
Redemption Interlude
Redemption (with Babes Wodumo)
Seasons (with Sjava & Reason)
Big Shot (with Travis Scott)
Pray For Me (with Kendrick Lamar)


In [0]:
# Function to pull song data from API using an artist's album uri, for all
# the albums that I chose.
def albumSongs(uri, name, album):
  
  album_tracks = []
  
  # pull data on album tracks
  tracks = sp.album_tracks(uri) 

  # for each track, create dictionary for that specific tracks
  for song in tracks['items']:
    track = {} #
    track["artist"] = name
    track['album'] = album
    track['uri'] = song['uri']
    track['name'] = song['name']
    track['artists'] = []
    for artist in song['artists']:
      track['artists'].append(artist["name"])
    track['track_number'] = song['track_number']
    track['duration_ms'] = song['duration_ms']
    track['explicit'] = song['explicit']
    # append each track dictionary to the list
    album_tracks.append(track)

  return album_tracks

In [0]:
spotify_tracks = []
album_count = 0
for album in new_album_uris:
    spotify_tracks.extend(albumSongs(album, name, new_album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 

Album Black Panther The Album Music From And Inspired By songs has been added to spotify_tracks
Album DAMN. COLLECTORS EDITION. songs has been added to spotify_tracks
Album untitled unmastered. songs has been added to spotify_tracks
Album To Pimp A Butterfly songs has been added to spotify_tracks
Album good kid, m.A.A.d city (Deluxe) songs has been added to spotify_tracks


In [0]:
df_kendrick = json_normalize(spotify_tracks)
df_kendrick

artist  ... explicit
0   Kendrick Lamar  ...     True
1   Kendrick Lamar  ...     True
2   Kendrick Lamar  ...     True
3   Kendrick Lamar  ...     True
4   Kendrick Lamar  ...     True
..             ...  ...      ...
64  Kendrick Lamar  ...     True
65  Kendrick Lamar  ...     True
66  Kendrick Lamar  ...    False
67  Kendrick Lamar  ...     True
68  Kendrick Lamar  ...     True

[69 rows x 8 columns]

In [0]:
df_kendrick['album'].value_counts()

good kid, m.A.A.d city (Deluxe)                       17
To Pimp A Butterfly                                   16
Black Panther The Album Music From And Inspired By    14
DAMN. COLLECTORS EDITION.                             14
untitled unmastered.                                   8
Name: album, dtype: int64

**Below here will be repeitions of above code for different hip hop artists.**

In [0]:
# Now I want to speed up doing this for multiple artists. For a diverse dataset
# I will get hip hop artists and and then another df with popartists.
# Unfortunately some songs have many artists, and the searched for artist does
# not always appear first in the artist search query, so I have to visually 
# inspect the result, instead of being able to loop through many names and 
# grabbing the first name for every result.
name = "J. Cole"

result = sp.search(name)
result['tracks']['items'][0]['artists'][0]

{'external_urls': {'spotify': 'https://open.spotify.com/artist/6l3HvQ5sa6mXTsMTB19rO5'},
 'href': 'https://api.spotify.com/v1/artists/6l3HvQ5sa6mXTsMTB19rO5',
 'id': '6l3HvQ5sa6mXTsMTB19rO5',
 'name': 'J. Cole',
 'type': 'artist',
 'uri': 'spotify:artist:6l3HvQ5sa6mXTsMTB19rO5'}

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][0]['uri']
sp_albums = sp.artist_albums(artist_uri, album_type='album')
album_names = []
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])

In [0]:
album_names

["Revenge Of The Dreamers III: Director's Cut",
 "Revenge Of The Dreamers III: Director's Cut",
 'Revenge Of The Dreamers III',
 'Revenge Of The Dreamers III',
 'KOD',
 'KOD',
 '4 Your Eyez Only',
 '4 Your Eyez Only',
 'Forest Hills Drive: Live from Fayetteville, NC',
 '2014 Forest Hills Drive',
 '2014 Forest Hills Drive',
 'The Blow Up',
 'Born Sinner',
 'Born Sinner',
 'Born Sinner',
 'Cole World: The Sideline Story',
 'Cole World: The Sideline Story',
 'The Come Up']

In [0]:
for album in album_uris:
  print (len(sp.album_tracks(album)["items"]))

30
30
18
18
12
12
10
10
13
13
13
11
16
21
16
16
16
18


**Unique albums that I want are at 0,4,6,9,11,12,15,17.**

In [0]:
indices = [0,4,6,9,11,12,15,17]
new_album_uris = [album_uris[x] for x in indices]
new_album_names = [album_names[x] for x in indices]
new_album_names

["Revenge Of The Dreamers III: Director's Cut",
 'KOD',
 '4 Your Eyez Only',
 '2014 Forest Hills Drive',
 'The Blow Up',
 'Born Sinner',
 'Cole World: The Sideline Story',
 'The Come Up']

In [0]:
spotify_tracks = []
album_count = 0
for album in new_album_uris:
    spotify_tracks.extend(albumSongs(album, name, new_album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 

Album Revenge Of The Dreamers III: Director's Cut songs has been added to spotify_tracks
Album KOD songs has been added to spotify_tracks
Album 4 Your Eyez Only songs has been added to spotify_tracks
Album 2014 Forest Hills Drive songs has been added to spotify_tracks
Album The Blow Up songs has been added to spotify_tracks
Album Born Sinner songs has been added to spotify_tracks
Album Cole World: The Sideline Story songs has been added to spotify_tracks
Album The Come Up songs has been added to spotify_tracks


In [0]:
df_jcole = json_normalize(spotify_tracks)
df_jcole

artist  ... explicit
0    J. Cole  ...     True
1    J. Cole  ...     True
2    J. Cole  ...     True
3    J. Cole  ...     True
4    J. Cole  ...     True
..       ...  ...      ...
121  J. Cole  ...    False
122  J. Cole  ...    False
123  J. Cole  ...    False
124  J. Cole  ...    False
125  J. Cole  ...    False

[126 rows x 8 columns]

In [0]:
name = "Drake"
result = sp.search(name)
result['tracks']['items'][0]['artists']

[{'external_urls': {'spotify': 'https://open.spotify.com/artist/1RyvyyTE3xzB2ZywiAwp0i'},
  'href': 'https://api.spotify.com/v1/artists/1RyvyyTE3xzB2ZywiAwp0i',
  'id': '1RyvyyTE3xzB2ZywiAwp0i',
  'name': 'Future',
  'type': 'artist',
  'uri': 'spotify:artist:1RyvyyTE3xzB2ZywiAwp0i'},
 {'external_urls': {'spotify': 'https://open.spotify.com/artist/3TVXtAsR1Inumwj472S9r4'},
  'href': 'https://api.spotify.com/v1/artists/3TVXtAsR1Inumwj472S9r4',
  'id': '3TVXtAsR1Inumwj472S9r4',
  'name': 'Drake',
  'type': 'artist',
  'uri': 'spotify:artist:3TVXtAsR1Inumwj472S9r4'}]

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][1]['uri']
sp_albums = sp.artist_albums(artist_uri, album_type='album')
album_names = []
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])

In [0]:
album_names

['Care Package',
 'Care Package',
 'So Far Gone',
 'So Far Gone',
 'Scorpion',
 'Scorpion',
 'More Life',
 'More Life',
 'Views',
 'Views',
 'Views',
 'Views',
 'What A Time To Be Alive',
 'What A Time To Be Alive',
 "If You're Reading This It's Too Late",
 "If You're Reading This It's Too Late",
 'Nothing Was The Same (Deluxe)',
 'Nothing Was The Same (Deluxe)',
 'Nothing Was The Same (Deluxe)',
 'Nothing Was The Same (Deluxe)']

In [0]:
for album in album_uris:
  print (len(sp.album_tracks(album)["items"]))

17
17
18
18
25
25
22
22
20
20
20
20
11
11
17
17
16
15
16
16


In [0]:
indices = [0,2,4,6,8,12,14,16]
new_album_uris = [album_uris[x] for x in indices]
new_album_names = [album_names[x] for x in indices]
new_album_names

['Care Package',
 'So Far Gone',
 'Scorpion',
 'More Life',
 'Views',
 'What A Time To Be Alive',
 "If You're Reading This It's Too Late",
 'Nothing Was The Same (Deluxe)']

In [0]:
spotify_tracks = []
album_count = 0
for album in new_album_uris:
    spotify_tracks.extend(albumSongs(album, name, new_album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 

Album Care Package songs has been added to spotify_tracks
Album So Far Gone songs has been added to spotify_tracks
Album Scorpion songs has been added to spotify_tracks
Album More Life songs has been added to spotify_tracks
Album Views songs has been added to spotify_tracks
Album What A Time To Be Alive songs has been added to spotify_tracks
Album If You're Reading This It's Too Late songs has been added to spotify_tracks
Album Nothing Was The Same (Deluxe) songs has been added to spotify_tracks


In [0]:
df_drake = json_normalize(spotify_tracks)
df_drake

artist                          album  ... duration_ms explicit
0    Drake                   Care Package  ...      253803     True
1    Drake                   Care Package  ...      240956     True
2    Drake                   Care Package  ...      235566     True
3    Drake                   Care Package  ...      281518     True
4    Drake                   Care Package  ...      353697     True
..     ...                            ...  ...         ...      ...
141  Drake  Nothing Was The Same (Deluxe)  ...      261866     True
142  Drake  Nothing Was The Same (Deluxe)  ...      433800     True
143  Drake  Nothing Was The Same (Deluxe)  ...      236360     True
144  Drake  Nothing Was The Same (Deluxe)  ...      271573     True
145  Drake  Nothing Was The Same (Deluxe)  ...      241000     True

[146 rows x 8 columns]

In [0]:
name = "Kanye West"
result = sp.search(name)
result['tracks']['items'][0]['artists']

[{'external_urls': {'spotify': 'https://open.spotify.com/artist/1cNDP5yjU5vjeR8qMf4grg'},
  'href': 'https://api.spotify.com/v1/artists/1cNDP5yjU5vjeR8qMf4grg',
  'id': '1cNDP5yjU5vjeR8qMf4grg',
  'name': 'YNW Melly',
  'type': 'artist',
  'uri': 'spotify:artist:1cNDP5yjU5vjeR8qMf4grg'},
 {'external_urls': {'spotify': 'https://open.spotify.com/artist/5K4W6rqBFWDnAN6FQUkS6x'},
  'href': 'https://api.spotify.com/v1/artists/5K4W6rqBFWDnAN6FQUkS6x',
  'id': '5K4W6rqBFWDnAN6FQUkS6x',
  'name': 'Kanye West',
  'type': 'artist',
  'uri': 'spotify:artist:5K4W6rqBFWDnAN6FQUkS6x'}]

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][1]["uri"]
sp_albums = sp.artist_albums(artist_uri, album_type='album')
album_names = []
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])

In [0]:
album_names

['JESUS IS KING',
 'KIDS SEE GHOSTS',
 'ye',
 'The Life Of Pablo',
 'The Life Of Pablo',
 'Yeezus',
 'Kanye West Presents Good Music Cruel Summer',
 'Watch The Throne (Deluxe)',
 'Watch The Throne (Deluxe)',
 'Watch The Throne (Deluxe)',
 'Watch The Throne (Deluxe)',
 'Watch The Throne',
 'Watch The Throne',
 'Watch The Throne',
 'Watch The Throne',
 'My Beautiful Dark Twisted Fantasy',
 'My Beautiful Dark Twisted Fantasy',
 '808s & Heartbreak',
 '808s & Heartbreak',
 'Graduation (French Limited Version)']

**0,1,2,3,5,6,7,15,17,19.**

In [0]:
for album in album_uris:
  print (len(sp.album_tracks(album)["items"]))

11
7
7
20
19
10
12
16
16
16
16
12
12
12
12
13
13
12
12
15


In [0]:
indices = [0,1,2,3,5,6,7,15,17,19]
new_album_uris = [album_uris[x] for x in indices]
new_album_names = [album_names[x] for x in indices]
new_album_names

['JESUS IS KING',
 'KIDS SEE GHOSTS',
 'ye',
 'The Life Of Pablo',
 'Yeezus',
 'Kanye West Presents Good Music Cruel Summer',
 'Watch The Throne (Deluxe)',
 'My Beautiful Dark Twisted Fantasy',
 '808s & Heartbreak',
 'Graduation (French Limited Version)']

In [0]:
spotify_tracks = []
album_count = 0
for album in new_album_uris:
    spotify_tracks.extend(albumSongs(album, name, new_album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 

Album JESUS IS KING songs has been added to spotify_tracks
Album KIDS SEE GHOSTS songs has been added to spotify_tracks
Album ye songs has been added to spotify_tracks
Album The Life Of Pablo songs has been added to spotify_tracks
Album Yeezus songs has been added to spotify_tracks
Album Kanye West Presents Good Music Cruel Summer songs has been added to spotify_tracks
Album Watch The Throne (Deluxe) songs has been added to spotify_tracks
Album My Beautiful Dark Twisted Fantasy songs has been added to spotify_tracks
Album 808s & Heartbreak songs has been added to spotify_tracks
Album Graduation (French Limited Version) songs has been added to spotify_tracks


In [0]:
df_kanye = json_normalize(spotify_tracks)
df_kanye

artist                                album  ... duration_ms explicit
0    Kanye West                        JESUS IS KING  ...      112239    False
1    Kanye West                        JESUS IS KING  ...      164619    False
2    Kanye West                        JESUS IS KING  ...      104591    False
3    Kanye West                        JESUS IS KING  ...      151873    False
4    Kanye West                        JESUS IS KING  ...      136350    False
..          ...                                  ...  ...         ...      ...
118  Kanye West  Graduation (French Limited Version)  ...      212826     True
119  Kanye West  Graduation (French Limited Version)  ...      203506     True
120  Kanye West  Graduation (French Limited Version)  ...      287413     True
121  Kanye West  Graduation (French Limited Version)  ...      276146    False
122  Kanye West  Graduation (French Limited Version)  ...      285880    False

[123 rows x 8 columns]

In [0]:
name = "Khalid"
result = sp.search(name)
result['tracks']['items'][0]['artists']

[{'external_urls': {'spotify': 'https://open.spotify.com/artist/6qqNVTkY8uBg9cP3Jd7DAH'},
  'href': 'https://api.spotify.com/v1/artists/6qqNVTkY8uBg9cP3Jd7DAH',
  'id': '6qqNVTkY8uBg9cP3Jd7DAH',
  'name': 'Billie Eilish',
  'type': 'artist',
  'uri': 'spotify:artist:6qqNVTkY8uBg9cP3Jd7DAH'},
 {'external_urls': {'spotify': 'https://open.spotify.com/artist/6LuN9FCkKOj5PcnpouEgny'},
  'href': 'https://api.spotify.com/v1/artists/6LuN9FCkKOj5PcnpouEgny',
  'id': '6LuN9FCkKOj5PcnpouEgny',
  'name': 'Khalid',
  'type': 'artist',
  'uri': 'spotify:artist:6LuN9FCkKOj5PcnpouEgny'}]

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][1]["uri"]
sp_albums = sp.artist_albums(artist_uri, album_type='album')
album_names = []
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])
album_names

['Free Spirit',
 'Suncity',
 'Suncity',
 'American Teen (Japan Version)',
 'American Teen']

In [0]:
indices = [0,1,4]
new_album_uris = [album_uris[x] for x in indices]
new_album_names = [album_names[x] for x in indices]
new_album_names

['Free Spirit', 'Suncity', 'American Teen']

In [0]:
spotify_tracks = []
album_count = 0
for album in new_album_uris:
    spotify_tracks.extend(albumSongs(album, name, new_album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 
df_khalid = json_normalize(spotify_tracks)
df_khalid

Album Free Spirit songs has been added to spotify_tracks
Album Suncity songs has been added to spotify_tracks
Album American Teen songs has been added to spotify_tracks


artist          album  ... duration_ms explicit
0   Khalid    Free Spirit  ...      213253    False
1   Khalid    Free Spirit  ...      231653    False
2   Khalid    Free Spirit  ...      163973    False
3   Khalid    Free Spirit  ...      229400    False
4   Khalid    Free Spirit  ...      197573    False
5   Khalid    Free Spirit  ...      215226    False
6   Khalid    Free Spirit  ...      165840    False
7   Khalid    Free Spirit  ...      173653    False
8   Khalid    Free Spirit  ...      277080    False
9   Khalid    Free Spirit  ...      177066    False
10  Khalid    Free Spirit  ...      182080    False
11  Khalid    Free Spirit  ...      184200    False
12  Khalid    Free Spirit  ...      199240    False
13  Khalid    Free Spirit  ...      229960    False
14  Khalid    Free Spirit  ...      177413    False
15  Khalid    Free Spirit  ...      213026    False
16  Khalid    Free Spirit  ...      211626    False
17  Khalid        Suncity  ...       55720    False
18  Khalid        Suncity  ...      270413    False
19  Khalid        Suncity  ...      209546    False
20  Khalid        Suncity  ...       81933     True
21  Khalid        Suncity  ...      235880    False
22  Khalid        Suncity  ...      229320    False
23  Khalid        Suncity  ...      189493    False
24  Khalid  American Teen  ...      250746    False
25  Khalid  American Teen  ...      202546    False
26  Khalid  American Teen  ...      219080    False
27  Khalid  American Teen  ...      244746    False
28  Khalid  American Teen  ...      206533    False
29  Khalid  American Teen  ...      199120    False
30  Khalid  American Teen  ...      228600    False
31  Khalid  American Teen  ...      204546     True
32  Khalid  American Teen  ...      167640    False
33  Khalid  American Teen  ...      207960    False
34  Khalid  American Teen  ...      241866    False
35  Khalid  American Teen  ...      257959    False
36  Khalid  American Teen  ...      276226    False
37  Khalid  American Teen  ...      207106    False
38  Khalid  American Teen  ...      170680    False

[39 rows x 8 columns]

In [0]:
# Concatenating all the data for hip hop artists together
df_hiphop = pd.concat([df_kendrick, df_jcole, df_drake, df_kanye, df_khalid])
df_hiphop = df_hiphop.set_index("uri")
df_hiphop

artist  ... explicit
uri                                                   ...         
spotify:track:5qN4HFkapdAOV94XPryVof  Kendrick Lamar  ...     True
spotify:track:3GCdLUSnKSMJhs4Tj6CV3s  Kendrick Lamar  ...     True
spotify:track:4LmAnpjlhWTahvRkYR8xJa  Kendrick Lamar  ...     True
spotify:track:5jyyPsIGM2yqkZN9R3TmvN  Kendrick Lamar  ...     True
spotify:track:7bUcBztfGqO7cSI2gMZeCI  Kendrick Lamar  ...     True
...                                              ...  ...      ...
spotify:track:4V3VshJLUTCIWa77YiAkvM          Khalid  ...    False
spotify:track:25YvGDl2zSE0pH8jrMZ6aY          Khalid  ...    False
spotify:track:6SWHTMywuTXLFlMQ75cAuu          Khalid  ...    False
spotify:track:1Ob94QesoxHSBrypyBbdfd          Khalid  ...    False
spotify:track:67kXYrMTQbS7MVbXyGyxHm          Khalid  ...    False

[503 rows x 7 columns]

In [0]:
df_hiphop["artist"].value_counts()

Drake             146
J. Cole           126
Kanye West        123
Kendrick Lamar     69
Khalid             39
Name: artist, dtype: int64

**Now I want the audio features of all the songs**

In [0]:
# Testing grabbing the audio features for one song
sp.audio_features(df_hiphop.index[0])

[{'acousticness': 0.625,
  'analysis_url': 'https://api.spotify.com/v1/audio-analysis/5qN4HFkapdAOV94XPryVof',
  'danceability': 0.618,
  'duration_ms': 130613,
  'energy': 0.582,
  'id': '5qN4HFkapdAOV94XPryVof',
  'instrumentalness': 3.84e-06,
  'key': 1,
  'liveness': 0.265,
  'loudness': -9.454,
  'mode': 1,
  'speechiness': 0.297,
  'tempo': 90.035,
  'time_signature': 4,
  'track_href': 'https://api.spotify.com/v1/tracks/5qN4HFkapdAOV94XPryVof',
  'type': 'audio_features',
  'uri': 'spotify:track:5qN4HFkapdAOV94XPryVof',
  'valence': 0.48}]

In [0]:
# Testing grabbing the popularity of a song
pop = sp.track(df_hiphop.index[0])
pop["popularity"]

57

In [0]:
# Function to grab the audio features for all the songs in the dataframe
def audio_features(df):
  tracks = []

  for uri in df.index:
    track_features = {}
    # pull audio features per track
    features = sp.audio_features(uri)
    
    # assign dictionary values for each audio feature
    track_features["uri"] = uri
    track_features['key'] = features[0]['key']
    track_features['mode'] = features[0]['mode']
    track_features['time_signature'] = features[0]['time_signature']
    track_features['acousticness'] = features[0]['acousticness']
    track_features['danceability'] = features[0]['danceability']
    track_features['energy'] = features[0]['energy']
    track_features['instrumentalness'] = features[0]['instrumentalness']
    track_features['liveness'] = features[0]['liveness']
    track_features['loudness'] = features[0]['loudness']
    track_features['speechiness'] = features[0]['speechiness']
    track_features['tempo'] = features[0]['tempo']
    track_features['valence'] = features[0]['valence']
    
    # grabbing popularity
    pop = sp.track(uri)
    track_features['popularity'] = pop['popularity']
    tracks.append(track_features)

  return tracks

In [0]:
tracks = audio_features(df_hiphop)
df_tracks = json_normalize(tracks)

retrying ...5secs
retrying ...5secs
retrying ...3secs
retrying ...1secs
retrying ...1secs
retrying ...1secs


In [0]:
df_tracks = df_tracks.set_index("uri")
df_tracks

key  mode  ...  valence  popularity
uri                                              ...                     
spotify:track:5qN4HFkapdAOV94XPryVof    1     1  ...    0.480          57
spotify:track:3GCdLUSnKSMJhs4Tj6CV3s    8     1  ...    0.552          78
spotify:track:4LmAnpjlhWTahvRkYR8xJa    2     1  ...    0.405          69
spotify:track:5jyyPsIGM2yqkZN9R3TmvN   11     0  ...    0.589          65
spotify:track:7bUcBztfGqO7cSI2gMZeCI    1     1  ...    0.847          59
...                                   ...   ...  ...      ...         ...
spotify:track:4V3VshJLUTCIWa77YiAkvM    1     1  ...    0.367          63
spotify:track:25YvGDl2zSE0pH8jrMZ6aY    9     0  ...    0.430          64
spotify:track:6SWHTMywuTXLFlMQ75cAuu    2     0  ...    0.663          62
spotify:track:1Ob94QesoxHSBrypyBbdfd    6     1  ...    0.489          68
spotify:track:67kXYrMTQbS7MVbXyGyxHm    6     0  ...    0.204          65

[503 rows x 13 columns]

In [0]:
# Merging tracks dataframe with audio features dataframe, I set uri as the index
# for both. uri is a unique song identifier, so I can merge cleanly on it
df_hiphop = df_hiphop.merge(df_tracks, left_index=True, 
                            right_index=True, how="outer")
# This will be the hiphop part of my data
df_hiphop["genre"] = "hiphop"
df_hiphop

artist  ...   genre
uri                                                   ...        
spotify:track:5qN4HFkapdAOV94XPryVof  Kendrick Lamar  ...  hiphop
spotify:track:3GCdLUSnKSMJhs4Tj6CV3s  Kendrick Lamar  ...  hiphop
spotify:track:4LmAnpjlhWTahvRkYR8xJa  Kendrick Lamar  ...  hiphop
spotify:track:5jyyPsIGM2yqkZN9R3TmvN  Kendrick Lamar  ...  hiphop
spotify:track:7bUcBztfGqO7cSI2gMZeCI  Kendrick Lamar  ...  hiphop
...                                              ...  ...     ...
spotify:track:4V3VshJLUTCIWa77YiAkvM          Khalid  ...  hiphop
spotify:track:25YvGDl2zSE0pH8jrMZ6aY          Khalid  ...  hiphop
spotify:track:6SWHTMywuTXLFlMQ75cAuu          Khalid  ...  hiphop
spotify:track:1Ob94QesoxHSBrypyBbdfd          Khalid  ...  hiphop
spotify:track:67kXYrMTQbS7MVbXyGyxHm          Khalid  ...  hiphop

[503 rows x 21 columns]

In [0]:
df_hiphop["album"].value_counts()

Revenge Of The Dreamers III: Director's Cut           30
Scorpion                                              25
More Life                                             22
Views                                                 20
The Life Of Pablo                                     20
The Come Up                                           18
So Far Gone                                           18
Free Spirit                                           17
good kid, m.A.A.d city (Deluxe)                       17
If You're Reading This It's Too Late                  17
Care Package                                          17
Nothing Was The Same (Deluxe)                         16
To Pimp A Butterfly                                   16
Watch The Throne (Deluxe)                             16
Born Sinner                                           16
Cole World: The Sideline Story                        16
Graduation (French Limited Version)                   15
American Teen                  

# Getting Pop Song Data from Spotify Web API

**I will just be repeating what I did with the hip hop data, with pop artists.**

In [0]:
name = "Ed Sheeran"
result = sp.search(name)
result['tracks']['items'][0]['artists']

[{'external_urls': {'spotify': 'https://open.spotify.com/artist/6eUKZXaKkcviH0Ku9w2n3V'},
  'href': 'https://api.spotify.com/v1/artists/6eUKZXaKkcviH0Ku9w2n3V',
  'id': '6eUKZXaKkcviH0Ku9w2n3V',
  'name': 'Ed Sheeran',
  'type': 'artist',
  'uri': 'spotify:artist:6eUKZXaKkcviH0Ku9w2n3V'},
 {'external_urls': {'spotify': 'https://open.spotify.com/artist/6LuN9FCkKOj5PcnpouEgny'},
  'href': 'https://api.spotify.com/v1/artists/6LuN9FCkKOj5PcnpouEgny',
  'id': '6LuN9FCkKOj5PcnpouEgny',
  'name': 'Khalid',
  'type': 'artist',
  'uri': 'spotify:artist:6LuN9FCkKOj5PcnpouEgny'}]

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][0]["uri"]
sp_albums = sp.artist_albums(artist_uri, album_type='album')
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])
album_names

['Free Spirit',
 'Suncity',
 'Suncity',
 'American Teen (Japan Version)',
 'American Teen',
 'No.6 Collaborations Project',
 'No.6 Collaborations Project',
 '÷ (Deluxe)',
 'x (Wembley Edition)',
 '5',
 'x (Deluxe Edition)',
 'x (Wembley Edition)',
 'x (Wembley Edition)',
 'x (Wembley Edition)',
 'Live at the Bedford',
 'Loose Change',
 '+',
 '+',
 '+',
 '+',
 '+',
 'No.5 Collaborations Project']

In [0]:
for album in album_uris:
  print (len(sp.album_tracks(album)["items"]))

15
15
16
23
32
16
24
24
24
6
8
18
16
16
12
12
8


**0,2,4,5,10,11,16**

In [0]:
indices = [0,2,4,5,10,11,16]
new_album_uris = [album_uris[x] for x in indices]
new_album_names = [album_names[x] for x in indices]
new_album_names

['Free Spirit',
 'Suncity',
 'American Teen',
 'No.6 Collaborations Project',
 'x (Deluxe Edition)',
 'x (Wembley Edition)',
 '+']

In [0]:
spotify_tracks = []
album_count = 0
for album in new_album_uris:
    spotify_tracks.extend(albumSongs(album, name, new_album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 
df_ed = json_normalize(spotify_tracks)
df_ed.head()

Album Free Spirit songs has been added to spotify_tracks
Album Suncity songs has been added to spotify_tracks
Album American Teen songs has been added to spotify_tracks
Album No.6 Collaborations Project songs has been added to spotify_tracks
Album x (Deluxe Edition) songs has been added to spotify_tracks
Album x (Wembley Edition) songs has been added to spotify_tracks
Album + songs has been added to spotify_tracks


artist        album  ... duration_ms explicit
0  Ed Sheeran  Free Spirit  ...      197866    False
1  Ed Sheeran  Free Spirit  ...      204466    False
2  Ed Sheeran  Free Spirit  ...      206186     True
3  Ed Sheeran  Free Spirit  ...      189733    False
4  Ed Sheeran  Free Spirit  ...      243266    False

[5 rows x 8 columns]

In [0]:
name = "Shawn Mendes"
result = sp.search(name)
result['tracks']['items'][0]['artists']

[{'external_urls': {'spotify': 'https://open.spotify.com/artist/7n2wHs1TKAczGzO7Dd2rGr'},
  'href': 'https://api.spotify.com/v1/artists/7n2wHs1TKAczGzO7Dd2rGr',
  'id': '7n2wHs1TKAczGzO7Dd2rGr',
  'name': 'Shawn Mendes',
  'type': 'artist',
  'uri': 'spotify:artist:7n2wHs1TKAczGzO7Dd2rGr'},
 {'external_urls': {'spotify': 'https://open.spotify.com/artist/4nDoRrQiYLoBzwC5BhVJzF'},
  'href': 'https://api.spotify.com/v1/artists/4nDoRrQiYLoBzwC5BhVJzF',
  'id': '4nDoRrQiYLoBzwC5BhVJzF',
  'name': 'Camila Cabello',
  'type': 'artist',
  'uri': 'spotify:artist:4nDoRrQiYLoBzwC5BhVJzF'}]

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][0]["uri"]
sp_albums = sp.artist_albums(artist_uri, album_type='album')
album_names = []
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])
album_names

['Shawn Mendes (Deluxe)',
 'Shawn Mendes',
 'MTV Unplugged',
 'Illuminate (Deluxe)',
 'Illuminate',
 'Live At Madison Square Garden',
 'Handwritten (Revisited)',
 'Handwritten',
 'Handwritten',
 'Handwritten (Deluxe)',
 'Handwritten (Deluxe)']

In [0]:
for album in album_uris:
  print (len(sp.album_tracks(album)["items"]))

16
14
11
16
13
12
16
13
12
16
15


**0,3,6**

In [0]:
indices = [0,3,6]
new_album_uris = [album_uris[x] for x in indices]
new_album_names = [album_names[x] for x in indices]
new_album_names

['Shawn Mendes (Deluxe)', 'Illuminate (Deluxe)', 'Handwritten (Revisited)']

In [0]:
spotify_tracks = []
album_count = 0
for album in new_album_uris:
    spotify_tracks.extend(albumSongs(album, name, new_album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 
df_shawn = json_normalize(spotify_tracks)
df_shawn.head()

Album Shawn Mendes (Deluxe) songs has been added to spotify_tracks
Album Illuminate (Deluxe) songs has been added to spotify_tracks
Album Handwritten (Revisited) songs has been added to spotify_tracks


artist                  album  ... duration_ms explicit
0  Shawn Mendes  Shawn Mendes (Deluxe)  ...      190799    False
1  Shawn Mendes  Shawn Mendes (Deluxe)  ...      200840    False
2  Shawn Mendes  Shawn Mendes (Deluxe)  ...      191466    False
3  Shawn Mendes  Shawn Mendes (Deluxe)  ...      211293    False
4  Shawn Mendes  Shawn Mendes (Deluxe)  ...      235760    False

[5 rows x 8 columns]

In [0]:
name = "Camila"
result = sp.search(name)
result['tracks']['items'][0]['artists']

[{'external_urls': {'spotify': 'https://open.spotify.com/artist/4nDoRrQiYLoBzwC5BhVJzF'},
  'href': 'https://api.spotify.com/v1/artists/4nDoRrQiYLoBzwC5BhVJzF',
  'id': '4nDoRrQiYLoBzwC5BhVJzF',
  'name': 'Camila Cabello',
  'type': 'artist',
  'uri': 'spotify:artist:4nDoRrQiYLoBzwC5BhVJzF'},
 {'external_urls': {'spotify': 'https://open.spotify.com/artist/4r63FhuTkUYltbVAg5TQnk'},
  'href': 'https://api.spotify.com/v1/artists/4r63FhuTkUYltbVAg5TQnk',
  'id': '4r63FhuTkUYltbVAg5TQnk',
  'name': 'DaBaby',
  'type': 'artist',
  'uri': 'spotify:artist:4r63FhuTkUYltbVAg5TQnk'}]

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][0]["uri"]
sp_albums = sp.artist_albums(artist_uri, album_type='album')
album_names = []
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])
album_names

['Romance', 'Romance', 'Camila']

In [0]:
indices = [0,2]
new_album_uris = [album_uris[x] for x in indices]
new_album_names = [album_names[x] for x in indices]

spotify_tracks = []
album_count = 0
for album in new_album_uris:
    spotify_tracks.extend(albumSongs(album, name, new_album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 
df_camila = json_normalize(spotify_tracks)
df_camila.head()

Album Romance songs has been added to spotify_tracks
Album Camila songs has been added to spotify_tracks


artist    album  ... duration_ms explicit
0  Camila  Romance  ...      219742    False
1  Camila  Romance  ...      194440    False
2  Camila  Romance  ...      200626    False
3  Camila  Romance  ...      170746    False
4  Camila  Romance  ...      190813    False

[5 rows x 8 columns]

In [0]:
name = "Taylor"
result = sp.search(name)
result['tracks']['items'][0]['artists']

[{'external_urls': {'spotify': 'https://open.spotify.com/artist/06HL4z0CvFAxyc27GXpf02'},
  'href': 'https://api.spotify.com/v1/artists/06HL4z0CvFAxyc27GXpf02',
  'id': '06HL4z0CvFAxyc27GXpf02',
  'name': 'Taylor Swift',
  'type': 'artist',
  'uri': 'spotify:artist:06HL4z0CvFAxyc27GXpf02'}]

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][0]["uri"]
sp_albums = sp.artist_albums(artist_uri, album_type='album')
album_names = []
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])
album_names

['Lover',
 'reputation',
 'reputation Stadium Tour Surprise Song Playlist',
 'reputation (Big Machine Radio Release Special)',
 'Taylor Swift Karaoke: 1989 (Deluxe)',
 '1989',
 '1989',
 'Taylor Swift Karaoke: 1989 (Deluxe)',
 'Taylor Swift Karaoke: 1989 (Deluxe Edition)',
 'Taylor Swift Karaoke: 1989',
 '1989 (Deluxe)',
 '1989 (Deluxe Edition)',
 '1989 (Deluxe Edition)',
 '1989 (Big Machine Radio Release Special)',
 '1989',
 '1989 (Deluxe)',
 'Red (Deluxe Edition)',
 'Red (Deluxe Edition)',
 'Red',
 'Red (Karaoke Version)']

In [0]:
for album in album_uris:
  print (len(sp.album_tracks(album)["items"]))

18
15
46
31
16
13
13
16
16
13
19
19
15
26
13
19
22
22
16
16


**0,1,10,16**

In [0]:
indices = [0,1,10,16]
new_album_uris = [album_uris[x] for x in indices]
new_album_names = [album_names[x] for x in indices]
new_album_names

['Lover', 'reputation', '1989 (Deluxe)', 'Red (Deluxe Edition)']

In [0]:
spotify_tracks = []
album_count = 0
for album in new_album_uris:
    spotify_tracks.extend(albumSongs(album, name, new_album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 
df_taylor = json_normalize(spotify_tracks)
df_taylor.head()

Album Lover songs has been added to spotify_tracks
Album reputation songs has been added to spotify_tracks
Album 1989 (Deluxe) songs has been added to spotify_tracks
Album Red (Deluxe Edition) songs has been added to spotify_tracks


artist  album  ... duration_ms explicit
0  Taylor  Lover  ...      170640    False
1  Taylor  Lover  ...      178426    False
2  Taylor  Lover  ...      221306    False
3  Taylor  Lover  ...      190360    False
4  Taylor  Lover  ...      211240    False

[5 rows x 8 columns]

In [0]:
name = "billie eilish"
result = sp.search(name)
result['tracks']['items'][0]['artists']

[{'external_urls': {'spotify': 'https://open.spotify.com/artist/6qqNVTkY8uBg9cP3Jd7DAH'},
  'href': 'https://api.spotify.com/v1/artists/6qqNVTkY8uBg9cP3Jd7DAH',
  'id': '6qqNVTkY8uBg9cP3Jd7DAH',
  'name': 'Billie Eilish',
  'type': 'artist',
  'uri': 'spotify:artist:6qqNVTkY8uBg9cP3Jd7DAH'}]

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][0]["uri"]
sp_albums = sp.artist_albums(artist_uri, album_type='album')
# Store artist's albums' names' and uris in separate lists
album_names = []
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])
album_names

['WHEN WE ALL FALL ASLEEP, WHERE DO WE GO?']

In [0]:
spotify_tracks = []
album_count = 0
for album in album_uris:
    spotify_tracks.extend(albumSongs(album, name, album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 
df_billie = json_normalize(spotify_tracks)
df_billie.head()

Album Lover songs has been added to spotify_tracks


artist  ... explicit
0  billie eilish  ...    False
1  billie eilish  ...    False
2  billie eilish  ...    False
3  billie eilish  ...    False
4  billie eilish  ...    False

[5 rows x 8 columns]

In [0]:
name = "Justin"
result = sp.search(name)
result['tracks']['items'][0]['artists']

[{'external_urls': {'spotify': 'https://open.spotify.com/artist/1uNFoZAHBGtllmzznpCI3s'},
  'href': 'https://api.spotify.com/v1/artists/1uNFoZAHBGtllmzznpCI3s',
  'id': '1uNFoZAHBGtllmzznpCI3s',
  'name': 'Justin Bieber',
  'type': 'artist',
  'uri': 'spotify:artist:1uNFoZAHBGtllmzznpCI3s'},
 {'external_urls': {'spotify': 'https://open.spotify.com/artist/0VRj0yCOv2FXJNP47XQnx5'},
  'href': 'https://api.spotify.com/v1/artists/0VRj0yCOv2FXJNP47XQnx5',
  'id': '0VRj0yCOv2FXJNP47XQnx5',
  'name': 'Quavo',
  'type': 'artist',
  'uri': 'spotify:artist:0VRj0yCOv2FXJNP47XQnx5'}]

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][0]["uri"]
sp_albums = sp.artist_albums(artist_uri, album_type='album')
album_names = []
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])
album_names

['Changes',
 'Purpose (Deluxe)',
 'Purpose (Deluxe)',
 'Journals',
 'Journals',
 'Believe Acoustic',
 'Believe Acoustic',
 'Believe',
 'Believe (Deluxe Edition)',
 'Believe (Deluxe Edition)',
 'Never Say Never - The Remixes',
 'Never Say Never - The Remixes',
 'Under The Mistletoe',
 'Under The Mistletoe',
 'Under The Mistletoe (Deluxe Edition)',
 'Under The Mistletoe (Deluxe Edition)',
 'Under The Mistletoe (Deluxe Edition)',
 'My World 2.0',
 'My Worlds (International Version)',
 'My Worlds (Oz Version)']

In [0]:
for album in album_uris:
  print (len(sp.album_tracks(album)["items"]))

17
19
19
16
15
11
11
13
17
16
7
7
11
11
16
15
15
10
18
19


**0,1,3,8,10,14,17**

In [0]:
indices = [0,1,3,8,10,14,17]
new_album_uris = [album_uris[x] for x in indices]
new_album_names = [album_names[x] for x in indices]
new_album_names

['Changes',
 'Purpose (Deluxe)',
 'Journals',
 'Believe (Deluxe Edition)',
 'Never Say Never - The Remixes',
 'Under The Mistletoe (Deluxe Edition)',
 'My World 2.0']

In [0]:
spotify_tracks = []
album_count = 0
for album in new_album_uris:
    spotify_tracks.extend(albumSongs(album, name, new_album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 
df_justin = json_normalize(spotify_tracks)
df_justin.head()

Album Changes songs has been added to spotify_tracks
Album Purpose (Deluxe) songs has been added to spotify_tracks
Album Journals songs has been added to spotify_tracks
Album Believe (Deluxe Edition) songs has been added to spotify_tracks
Album Never Say Never - The Remixes songs has been added to spotify_tracks
Album Under The Mistletoe (Deluxe Edition) songs has been added to spotify_tracks
Album My World 2.0 songs has been added to spotify_tracks


artist    album  ... duration_ms explicit
0  Justin  Changes  ...      136906    False
1  Justin  Changes  ...      168200    False
2  Justin  Changes  ...      200773    False
3  Justin  Changes  ...      212866    False
4  Justin  Changes  ...      208520    False

[5 rows x 8 columns]

In [0]:
name = "Ariana"
result = sp.search(name)
result['tracks']['items'][0]['artists']

[{'external_urls': {'spotify': 'https://open.spotify.com/artist/66CXWjxzNUsdJxJ2JdwvnR'},
  'href': 'https://api.spotify.com/v1/artists/66CXWjxzNUsdJxJ2JdwvnR',
  'id': '66CXWjxzNUsdJxJ2JdwvnR',
  'name': 'Ariana Grande',
  'type': 'artist',
  'uri': 'spotify:artist:66CXWjxzNUsdJxJ2JdwvnR'}]

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][0]["uri"]
sp_albums = sp.artist_albums(artist_uri, album_type='album')
album_names = []
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])
album_names

['k bye for now (swt live)',
 'thank u, next',
 'Sweetener',
 'Sweetener',
 'thank u, next',
 'Christmas & Chill (Japan Version)',
 'Dangerous Woman',
 'Dangerous Woman',
 'Dangerous Woman',
 'Dangerous Woman',
 'Dangerous Woman',
 'Dangerous Woman',
 'Dangerous Woman',
 'My Everything (Italian Edition)',
 'The Remix',
 'My Everything (Deluxe)',
 'My Everything',
 'My Everything',
 'My Everything (Deluxe)',
 'My Everything']

In [0]:
for album in album_uris:
  print (len(sp.album_tracks(album)["items"]))

32
12
15
15
12
7
16
15
11
15
15
11
15
18
15
18
18
15
15
12


**1,2,5,14,15**

In [0]:
indices = [1,2,5,14,15]
new_album_uris = [album_uris[x] for x in indices]
new_album_names = [album_names[x] for x in indices]
new_album_names

['thank u, next',
 'Sweetener',
 'Christmas & Chill (Japan Version)',
 'The Remix',
 'My Everything (Deluxe)']

In [0]:
spotify_tracks = []
album_count = 0
for album in new_album_uris:
    spotify_tracks.extend(albumSongs(album, name, new_album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 
df_ariana = json_normalize(spotify_tracks)
df_ariana.head()

Album thank u, next songs has been added to spotify_tracks
Album Sweetener songs has been added to spotify_tracks
Album Christmas & Chill (Japan Version) songs has been added to spotify_tracks
Album The Remix songs has been added to spotify_tracks
Album My Everything (Deluxe) songs has been added to spotify_tracks


artist          album  ... duration_ms explicit
0  Ariana  thank u, next  ...      212266     True
1  Ariana  thank u, next  ...      171573     True
2  Ariana  thank u, next  ...      182000    False
3  Ariana  thank u, next  ...      216893     True
4  Ariana  thank u, next  ...      208880     True

[5 rows x 8 columns]

In [0]:
name = "ZAYN"
result = sp.search(name)
result['tracks']['items'][0]['artists']

[{'external_urls': {'spotify': 'https://open.spotify.com/artist/3KwmxIhSe9UTSEF37kwngR'},
  'href': 'https://api.spotify.com/v1/artists/3KwmxIhSe9UTSEF37kwngR',
  'id': '3KwmxIhSe9UTSEF37kwngR',
  'name': 'SHAED',
  'type': 'artist',
  'uri': 'spotify:artist:3KwmxIhSe9UTSEF37kwngR'},
 {'external_urls': {'spotify': 'https://open.spotify.com/artist/5ZsFI1h6hIdQRw2ti0hz81'},
  'href': 'https://api.spotify.com/v1/artists/5ZsFI1h6hIdQRw2ti0hz81',
  'id': '5ZsFI1h6hIdQRw2ti0hz81',
  'name': 'ZAYN',
  'type': 'artist',
  'uri': 'spotify:artist:5ZsFI1h6hIdQRw2ti0hz81'}]

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][1]["uri"]
sp_albums = sp.artist_albums(artist_uri, album_type='album')
album_names = []
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])
album_names

['Icarus Falls',
 'Icarus Falls',
 'Icarus Falls',
 'LIKE I WOULD (The Remixes)',
 'Mind Of Mine (Japan Version)',
 'Mind Of Mine (Deluxe Edition)',
 'Mind Of Mine (Deluxe Edition)']

In [0]:
indices = [0,3,5]
new_album_uris = [album_uris[x] for x in indices]
new_album_names = [album_names[x] for x in indices]
new_album_names

['Icarus Falls', 'LIKE I WOULD (The Remixes)', 'Mind Of Mine (Deluxe Edition)']

In [0]:
spotify_tracks = []
album_count = 0
for album in new_album_uris:
    spotify_tracks.extend(albumSongs(album, name, new_album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 
df_zayn = json_normalize(spotify_tracks)
df_zayn.head()

Album Icarus Falls songs has been added to spotify_tracks
Album LIKE I WOULD (The Remixes) songs has been added to spotify_tracks
Album Mind Of Mine (Deluxe Edition) songs has been added to spotify_tracks


artist         album  ... duration_ms explicit
0   ZAYN  Icarus Falls  ...      185373    False
1   ZAYN  Icarus Falls  ...      193186    False
2   ZAYN  Icarus Falls  ...      195466    False
3   ZAYN  Icarus Falls  ...      232520    False
4   ZAYN  Icarus Falls  ...      190120    False

[5 rows x 8 columns]

In [0]:
name = "Halsey"
result = sp.search(name)
result['tracks']['items'][0]['artists']

[{'external_urls': {'spotify': 'https://open.spotify.com/artist/26VFTg2z8YR0cCuwLzESi2'},
  'href': 'https://api.spotify.com/v1/artists/26VFTg2z8YR0cCuwLzESi2',
  'id': '26VFTg2z8YR0cCuwLzESi2',
  'name': 'Halsey',
  'type': 'artist',
  'uri': 'spotify:artist:26VFTg2z8YR0cCuwLzESi2'}]

In [0]:
artist_uri = result['tracks']['items'][0]['artists'][0]["uri"]
sp_albums = sp.artist_albums(artist_uri, album_type='album')
album_names = []
album_uris = []
for i in range(len(sp_albums['items'])):
    album_names.append(sp_albums['items'][i]['name'])
    album_uris.append(sp_albums['items'][i]['uri'])
album_names

['Manic',
 'hopeless fountain kingdom (Deluxe)',
 'hopeless fountain kingdom',
 'hopeless fountain kingdom',
 'hopeless fountain kingdom',
 'hopeless fountain kingdom (Deluxe)',
 'hopeless fountain kingdom',
 'BADLANDS (Deluxe)',
 'BADLANDS (Deluxe)',
 'BADLANDS',
 'BADLANDS']

In [0]:
for album in album_uris:
  print (len(sp.album_tracks(album)["items"]))

16
16
16
13
13
16
13
16
16
11
11


In [0]:
indices = [0,2,7]
new_album_uris = [album_uris[x] for x in indices]
new_album_names = [album_names[x] for x in indices]
new_album_names

['Manic', 'hopeless fountain kingdom', 'BADLANDS (Deluxe)']

In [0]:
spotify_tracks = []
album_count = 0
for album in new_album_uris:
    spotify_tracks.extend(albumSongs(album, name, new_album_names[album_count]))
    print("Album " + str(new_album_names[album_count]) + " songs has been added to spotify_tracks")
    album_count += 1 
df_halsey = json_normalize(spotify_tracks)
df_halsey.head()

Album Manic songs has been added to spotify_tracks
Album hopeless fountain kingdom songs has been added to spotify_tracks
Album BADLANDS (Deluxe) songs has been added to spotify_tracks


artist  album  ... duration_ms explicit
0  Halsey  Manic  ...      186187     True
1  Halsey  Manic  ...      234410    False
2  Halsey  Manic  ...      181805    False
3  Halsey  Manic  ...      205473     True
4  Halsey  Manic  ...      167904    False

[5 rows x 8 columns]

In [0]:
df_pop = pd.concat([df_ed, df_ariana, df_justin, df_billie, df_taylor, 
                       df_camila, df_shawn, df_zayn, df_halsey])
df_pop = df_pop.set_index("uri")
df_pop

artist  ... explicit
uri                                               ...         
spotify:track:70eFcWOvlMObDhURTqT4Fv  Ed Sheeran  ...    False
spotify:track:4vUmTMuQqjdnvlZmAH61Qk  Ed Sheeran  ...    False
spotify:track:4wuCQX7JvAZLlrcmH4AeZF  Ed Sheeran  ...     True
spotify:track:1AI7UPw3fgwAFkvAlZWhE0  Ed Sheeran  ...    False
spotify:track:0VsGaRXR5WAzpu51unJTis  Ed Sheeran  ...    False
...                                          ...  ...      ...
spotify:track:1r1WtSrPbAOYHE7zwq4eqK      Halsey  ...    False
spotify:track:2IO7yf562c1zLzpanal1DT      Halsey  ...     True
spotify:track:3SKzTIl9mW0DadTys9SFo3      Halsey  ...    False
spotify:track:5x2XIAdvFxWCwIOMNkbWUj      Halsey  ...     True
spotify:track:36dfGEKTGHxLmdNIYr0VTi      Halsey  ...    False

[545 rows x 7 columns]

In [0]:
# Using same audio_features function used on hiphop dataset
tracks = audio_features(df_pop)
df_tracks = json_normalize(tracks)

retrying ...1secs
retrying ...5secs
retrying ...2secs
retrying ...4secs
retrying ...3secs
retrying ...5secs
retrying ...4secs


In [0]:
df_tracks.head()

uri  key  ...  valence  popularity
0  spotify:track:70eFcWOvlMObDhURTqT4Fv    5  ...    0.548          80
1  spotify:track:4vUmTMuQqjdnvlZmAH61Qk    9  ...    0.668          86
2  spotify:track:4wuCQX7JvAZLlrcmH4AeZF    4  ...    0.607          69
3  spotify:track:1AI7UPw3fgwAFkvAlZWhE0    8  ...    0.605          74
4  spotify:track:0VsGaRXR5WAzpu51unJTis    2  ...    0.237          71

[5 rows x 14 columns]

In [0]:
df_tracks = df_tracks.set_index("uri")
df_pop = df_pop.merge(df_tracks, left_index=True, right_index=True, how="outer")
df_pop["genre"] = "pop"
df_pop.head()

artist        album  ... popularity genre
uri                                                            ...                 
spotify:track:70eFcWOvlMObDhURTqT4Fv  Ed Sheeran  Free Spirit  ...         80   pop
spotify:track:4vUmTMuQqjdnvlZmAH61Qk  Ed Sheeran  Free Spirit  ...         86   pop
spotify:track:4wuCQX7JvAZLlrcmH4AeZF  Ed Sheeran  Free Spirit  ...         69   pop
spotify:track:1AI7UPw3fgwAFkvAlZWhE0  Ed Sheeran  Free Spirit  ...         74   pop
spotify:track:0VsGaRXR5WAzpu51unJTis  Ed Sheeran  Free Spirit  ...         71   pop

[5 rows x 21 columns]

**The song names currently include features, which is not useful since that data is already included in artists. Also, I need a clean song name to scrape the lyrics of the songs.**

In [0]:
# Make new column with just clean song names using regex
import re
df_pop["name"] = df_pop["name"].str.replace(r' \([^)]*\)', "")
df_hiphop["name"] = df_hiphop["name"].str.replace(r' \([^)]*\)', "")
df_pop.head()

artist        album  ... popularity genre
uri                                                            ...                 
spotify:track:70eFcWOvlMObDhURTqT4Fv  Ed Sheeran  Free Spirit  ...         80   pop
spotify:track:4vUmTMuQqjdnvlZmAH61Qk  Ed Sheeran  Free Spirit  ...         86   pop
spotify:track:4wuCQX7JvAZLlrcmH4AeZF  Ed Sheeran  Free Spirit  ...         69   pop
spotify:track:1AI7UPw3fgwAFkvAlZWhE0  Ed Sheeran  Free Spirit  ...         74   pop
spotify:track:0VsGaRXR5WAzpu51unJTis  Ed Sheeran  Free Spirit  ...         71   pop

[5 rows x 21 columns]

In [0]:
# Save to csv so I dont have to repeat everything above
df_hiphop.to_csv("/content/drive/My Drive/df_hiphop.csv")
df_pop.to_csv("/content/drive/My Drive/df_pop.csv")

# Scraping Song Lyrics from MetroLyrics

In [0]:
# Concatenating the hiphop and pop dataframes
df_songs = pd.concat([df_hiphop, df_pop])
df_songs

artist  ...   genre
uri                                                   ...        
spotify:track:5qN4HFkapdAOV94XPryVof  Kendrick Lamar  ...  hiphop
spotify:track:3GCdLUSnKSMJhs4Tj6CV3s  Kendrick Lamar  ...  hiphop
spotify:track:4LmAnpjlhWTahvRkYR8xJa  Kendrick Lamar  ...  hiphop
spotify:track:5jyyPsIGM2yqkZN9R3TmvN  Kendrick Lamar  ...  hiphop
spotify:track:7bUcBztfGqO7cSI2gMZeCI  Kendrick Lamar  ...  hiphop
...                                              ...  ...     ...
spotify:track:1r1WtSrPbAOYHE7zwq4eqK          Halsey  ...     pop
spotify:track:2IO7yf562c1zLzpanal1DT          Halsey  ...     pop
spotify:track:3SKzTIl9mW0DadTys9SFo3          Halsey  ...     pop
spotify:track:5x2XIAdvFxWCwIOMNkbWUj          Halsey  ...     pop
spotify:track:36dfGEKTGHxLmdNIYr0VTi          Halsey  ...     pop

[1048 rows x 21 columns]

In [0]:
df_songs["artist"].value_counts()

Drake             146
J. Cole           126
Kanye West        123
Ed Sheeran        113
Justin            102
Taylor             74
Kendrick Lamar     69
Ariana             67
ZAYN               54
Halsey             48
Shawn Mendes       48
Khalid             39
Camila             25
billie eilish      14
Name: artist, dtype: int64

**I need the complete artist name to web scrape the song lyrics.**

In [0]:
# Mapping incomplete names to their complete names
dictionary = {
    "Kanye": "Kanye West",
    "Justin": "Justin Bieber",
    "Taylor": "Taylor Swift",
    "Ariana": "Ariana Grande",
    "Camila": "Camila Cabello"
}
df_songs["artist"] = df_songs["artist"].apply(
    lambda x: dictionary[x] if x in dictionary else x)

df_songs["artist"].value_counts()

Drake             146
J. Cole           126
Kanye West        123
Ed Sheeran        113
Justin Bieber     102
Taylor Swift       74
Kendrick Lamar     69
Ariana Grande      67
ZAYN               54
Halsey             48
Shawn Mendes       48
Khalid             39
Camila Cabello     25
billie eilish      14
Name: artist, dtype: int64

In [0]:
# Testing web scraping off of metro lyrics
import requests
from bs4 import BeautifulSoup
import re
import string

url = "https://www.metrolyrics.com/i-dont-care-lyrics-ed-sheeran.html"

content = requests.get(url)
soup = BeautifulSoup(content.content, 'html.parser')

# Metrolyrics has nice verse tag, but each song has multiple verses
a = soup.findAll("p", class_="verse")
 
# Formatting the lyrics text, will explain in the loop
c = " ".join([b.text for b in a])
d = re.sub(r'\[.*?\]', "", c).strip().replace("\n", " \n ").lower()
e = d.translate(str.maketrans('', '', string.punctuation))
f = re.sub('[()]', '', e)
g = re.sub(r' +', " ", f)
h = re.sub(r' +', " ", g.replace("\n", " "))
h

'im at a party i dont wanna be at and i dont ever wear a suit and tie yeah wonderin if i could sneak out the back nobodys even lookin me in my eyes can you take my hand finish my drink say shall we dance hell yeah you know i love ya did i ever tell ya you make it better like that dont think i fit in at this party everyones got so much to say yeah i always feel like im nobody mmm who wants to fit in anyway cause i dont care when im with my baby yeah all the bad things disappear and youre making me feel like maybe i am somebody i can deal with the bad nights when im with my baby yeah ooh ooh ooh ooh ooh ooh cause i dont care as long as you just hold me near you can take me anywhere and youre making me feel like im loved by somebody i can deal with the bad nights when im with my baby yeah ooh ooh ooh ooh ooh ooh we at a party we dont wanna be at tryna talk but we cant hear ourselves press your lips id rather kiss em right back with all these people all around im crippled with anxiety but 

In [0]:
# Testing generating a url from the dataframe columns in the format needed
# An example of an url is above. Urls don't have any punctuation except -
songs = zip(df_songs["artist"], df_songs["name"])
url = "https://www.metrolyrics.com/"
a = list(songs)
# Making a list in the right order, including the lyrics string in between
b = a[0][1].split() + ["lyrics"] + a[0][0].split()
# Removing punctuation from each string in list
c = [c.translate(str.maketrans('', '', string.punctuation)) for c in b]
# Joining the list with a - between each string
d = "-".join(c).lower()
d += ".html"
url += d
url

'https://www.metrolyrics.com/black-panther-lyrics-kendrick-lamar.html'

In [0]:
songs = zip(df_songs["artist"], df_songs["name"])

In [0]:
# Web scraping the lyrics for songs off of metrolyrics.com
import time

song_lyrics = []
for song in songs:
  url = "https://www.metrolyrics.com/"
  # Dictionary for the song lyrics
  lyrics_dict = {}

  try:
    a = list(song)
    b = list(a)[1].split() + ["lyrics"] + list(a)[0].split()
    c = [c.translate(str.maketrans('', '', string.punctuation)) for c in b]
    d = "-".join(c).lower()
    d += ".html"
    url += d

    content = requests.get(url)
    soup = BeautifulSoup(content.content, 'html.parser')

    # Genius has nice class lable "lyrics" for div, which makes scraping easy
    a = soup.findAll("p", class_="verse")
    # Song lyrics are text separated by html tags, so I join them with a space
    c = " ".join([b.text for b in a])
    # Removing verse labels, and adding spaces to \n for even spacing
    d = re.sub(r'\[.*?\]', "", c).strip().replace("\n", " \n ").lower()
    # Removing punctuation
    e = d.translate(str.maketrans('', '', string.punctuation))
    # Removing parentheses around any song lyrics
    f = re.sub('[()]', '', e)
    # Replacing multiple spaces with just one, this will be chain_lyrics
    g = re.sub(r' +', " ", f)
    # This will be the clean lyrics without \n
    h = re.sub(r' +', " ", g.replace("\n", " "))
    lyrics_dict["chain_lyrics"] = g
    lyrics_dict["lyrics"] = h

    song_lyrics.append(lyrics_dict)

  # Try except in case there is an error in the scraping, append empty dict
  except Exception as e:
    song_lyrics.append({})
  
  time.sleep(0.5)

In [0]:
df_lyrics = json_normalize(song_lyrics)
df_lyrics[df_lyrics["lyrics"]!=""].count()

chain_lyrics    831
lyrics          831
dtype: int64

**So I managed to get 831 total songs using that url generation, which is good enough. From other analysis I've noted that many songs aren't even on metrolyrics possible due to being not very well known, and other songs were live versions of songs, of which the url generation wouldn't work. There are other reasons why it could've failed. I will stick with the 831 songs.**

In [0]:
df_songs.reset_index().head()

index          artist  ...   genre duration_s
0      0  Kendrick Lamar  ...  hiphop    130.613
1      1  Kendrick Lamar  ...  hiphop    232.186
2      2  Kendrick Lamar  ...  hiphop    267.426
3      3  Kendrick Lamar  ...  hiphop    238.893
4      4  Kendrick Lamar  ...  hiphop    180.893

[5 rows x 20 columns]

In [0]:
df_complete_lyrics = df_songs.reset_index().iloc[:,1:].merge(
    df_lyrics, left_index=True, right_index=True, how="outer")
df_complete_lyrics

artist  ...                                             lyrics
0     Kendrick Lamar  ...  why i go easy know why i go easy wait king of ...
1     Kendrick Lamar  ...  love lets talk about love is it anything and e...
2     Kendrick Lamar  ...                                                   
3     Kendrick Lamar  ...                                                   
4     Kendrick Lamar  ...                                                   
...              ...  ...                                                ...
1043          Halsey  ...  keep on haunting keep on haunting me keep on h...
1044          Halsey  ...  are you insane like me been in pain like me bo...
1045          Halsey  ...  they send me away to find them a fortune a che...
1046          Halsey  ...  he says oh baby girl you know were gonna be le...
1047          Halsey  ...  i keep a close watch on this heart of mine i k...

[1048 rows x 21 columns]

In [0]:
# For this dataset, I'm not interested in songs for which I don't
# have the lyrics, so I drop those rows.
df_complete_lyrics = df_complete_lyrics[df_complete_lyrics["lyrics"]!=""]
df_complete_lyrics

artist  ...                                             lyrics
0     Kendrick Lamar  ...  why i go easy know why i go easy wait king of ...
1     Kendrick Lamar  ...  love lets talk about love is it anything and e...
5     Kendrick Lamar  ...  everybody put three fingers in the air the sky...
12    Kendrick Lamar  ...  wakanda welcome big shot hol up wait peanut bu...
13    Kendrick Lamar  ...  im always ready for a war again go down that r...
...              ...  ...                                                ...
1043          Halsey  ...  keep on haunting keep on haunting me keep on h...
1044          Halsey  ...  are you insane like me been in pain like me bo...
1045          Halsey  ...  they send me away to find them a fortune a che...
1046          Halsey  ...  he says oh baby girl you know were gonna be le...
1047          Halsey  ...  i keep a close watch on this heart of mine i k...

[831 rows x 21 columns]

In [0]:
df_complete_lyrics = df_complete_lyrics.reset_index().iloc[:,1:]
df_complete_lyrics

artist  ...                                             lyrics
0    Kendrick Lamar  ...  why i go easy know why i go easy wait king of ...
1    Kendrick Lamar  ...  love lets talk about love is it anything and e...
2    Kendrick Lamar  ...  everybody put three fingers in the air the sky...
3    Kendrick Lamar  ...  wakanda welcome big shot hol up wait peanut bu...
4    Kendrick Lamar  ...  im always ready for a war again go down that r...
..              ...  ...                                                ...
826          Halsey  ...  keep on haunting keep on haunting me keep on h...
827          Halsey  ...  are you insane like me been in pain like me bo...
828          Halsey  ...  they send me away to find them a fortune a che...
829          Halsey  ...  he says oh baby girl you know were gonna be le...
830          Halsey  ...  i keep a close watch on this heart of mine i k...

[831 rows x 21 columns]

**This dataframe is for textual features, so I can drop the audio features.**

In [0]:
df_complete_lyrics = df_complete_lyrics.iloc[:,[0,1,2,3,17,19,20]]
df_complete_lyrics

artist  ...                                             lyrics
0    Kendrick Lamar  ...  why i go easy know why i go easy wait king of ...
1    Kendrick Lamar  ...  love lets talk about love is it anything and e...
2    Kendrick Lamar  ...  everybody put three fingers in the air the sky...
3    Kendrick Lamar  ...  wakanda welcome big shot hol up wait peanut bu...
4    Kendrick Lamar  ...  im always ready for a war again go down that r...
..              ...  ...                                                ...
826          Halsey  ...  keep on haunting keep on haunting me keep on h...
827          Halsey  ...  are you insane like me been in pain like me bo...
828          Halsey  ...  they send me away to find them a fortune a che...
829          Halsey  ...  he says oh baby girl you know were gonna be le...
830          Halsey  ...  i keep a close watch on this heart of mine i k...

[831 rows x 7 columns]

In [0]:
df_complete_lyrics.to_csv("/content/drive/My Drive/df_complete_lyrics.csv")

# Getting Sentiment Analysis from Google Cloud Natural Language API.

In [0]:
APIKEY = 

**The json response object is very simple, with just a polarity and a magnitude. Don't need to practice or inspect the response.**

In [0]:
# Grabbing sentiment analysis or lyrics from Google Cloud Natural Language API.
# I think Google likely will perform better than perhaps a python library
from googleapiclient.discovery import build
lservice = build('language', 'v1beta1', developerKey=APIKEY)

sentiments = []
for lyric in df_complete_lyrics["lyrics"]:
  sentiment= {}
  response = lservice.documents().analyzeSentiment(
    body={
      'document': {
         'type': 'PLAIN_TEXT',
         'content': lyric
      }
    }).execute()
  sentiment["polarity"] = response['documentSentiment']['polarity']
  sentiment["magnitude"] = response['documentSentiment']['magnitude']
  sentiments.append(sentiment)

In [0]:
df_sentiment = json_normalize(sentiments)
df_sentiment

polarity  magnitude
0        -1.0        1.2
1        -1.0        1.3
2        -1.0        2.2
3        -1.0        1.3
4        -1.0        1.1
..        ...        ...
826      -1.0        1.4
827      -1.0        1.0
828      -1.0        1.7
829       0.2        1.1
830       1.0        0.6

[831 rows x 2 columns]

In [0]:
# Values are passed in order of original dataframe, so I can reset index of 
# original and merge the two together. Resetting index creates extra column
# that I drop
df_songs_sentiment = df_complete_lyrics.reset_index().iloc[:,1:].merge(
    df_sentiment, left_index=True, right_index=True, how="outer")
df_songs_sentiment

artist  ... magnitude
0    Kendrick Lamar  ...       1.2
1    Kendrick Lamar  ...       1.3
2    Kendrick Lamar  ...       2.2
3    Kendrick Lamar  ...       1.3
4    Kendrick Lamar  ...       1.1
..              ...  ...       ...
826          Halsey  ...       1.4
827          Halsey  ...       1.0
828          Halsey  ...       1.7
829          Halsey  ...       1.1
830          Halsey  ...       0.6

[831 rows x 9 columns]

# Getting Personality Analysis from IBM Watson API Personality Insights

In [0]:
pip install --upgrade "ibm-watson>=4.3.0"

     |████████████████████████████████| 368kB 4.9MB/s 
     |████████████████████████████████| 204kB 50.0MB/s 
  Created wheel for ibm-watson: filename=ibm_watson-4.3.0-cp36-none-any.whl size=354170 sha256=2a1b2cb668535494ede5fb30228c6634d739f23c324b3ecaeb23818c8b4bfc93
  Stored in directory: /root/.cache/pip/wheels/b7/10/b4/509307276e9d558a6b06d8e69f8faf6cfbdf2d1b83abb6b8d4
  Created wheel for ibm-cloud-sdk-core: filename=ibm_cloud_sdk_core-1.5.1-cp36-none-any.whl size=44491 sha256=ed34176e32474f48e15e23fb2f66e1272e8bbc9c12e668615fb69b8abeb3acea
  Stored in directory: /root/.cache/pip/wheels/6a/42/50/f96888116b329578304f9dda4693cef6f3e76e18272d22cb6c
Successfully built ibm-watson ibm-cloud-sdk-core


In [0]:
APIKEY = 

In [0]:
# Practicing getting one result for one text "document"
from ibm_watson import PersonalityInsightsV3
from ibm_cloud_sdk_core.authenticators import IAMAuthenticator
# Authenticator object
authenticator = IAMAuthenticator(APIKEY)
personality_insights = PersonalityInsightsV3(
    version= '2017-10-13',
    authenticator=authenticator
)
# Dallas service url, closest to SLO
personality_insights.set_service_url(
    'https://api.us-south.personality-insights.watson.cloud.ibm.com')

profile = personality_insights.profile(
        df_songs_sentiment.loc[0]["lyrics"],
        accept='application/json',
        content_type='text/html',
).get_result()

**The response json contains 5 categories: Agreeableness, Conscientiousness, Extraversion, Emotional range, and Openness. Each category has a list of personality types for which a percentile is calcualted. I will look at the result for one text document and see which personality types I want to extract (whatever ones I think are good predictors.**

In [0]:
# Inspecting the results for each of the big 5 categories
profile["personality"][0]["children"]

[{'category': 'personality',
  'name': 'Adventurousness',
  'percentile': 0.7713670210526818,
  'significant': True,
  'trait_id': 'facet_adventurousness'},
 {'category': 'personality',
  'name': 'Artistic interests',
  'percentile': 0.9999827901171386,
  'significant': True,
  'trait_id': 'facet_artistic_interests'},
 {'category': 'personality',
  'name': 'Emotionality',
  'percentile': 0.9315941301830221,
  'significant': True,
  'trait_id': 'facet_emotionality'},
 {'category': 'personality',
  'name': 'Imagination',
  'percentile': 0.9983841275049014,
  'significant': True,
  'trait_id': 'facet_imagination'},
 {'category': 'personality',
  'name': 'Intellect',
  'percentile': 0.9989852735576126,
  'significant': True,
  'trait_id': 'facet_intellect'},
 {'category': 'personality',
  'name': 'Authority-challenging',
  'percentile': 0.9952214483173324,
  'significant': True,
  'trait_id': 'facet_liberalism'}]

**I want artistic, emotionality, and imagination, which are at
profile["personality"][0]["children"] items 1,2,and 3.**

In [0]:
profile["personality"][1]["children"]

[{'category': 'personality',
  'name': 'Achievement striving',
  'percentile': 0.33591482594587113,
  'significant': True,
  'trait_id': 'facet_achievement_striving'},
 {'category': 'personality',
  'name': 'Cautiousness',
  'percentile': 0.48078344770072584,
  'significant': True,
  'trait_id': 'facet_cautiousness'},
 {'category': 'personality',
  'name': 'Dutifulness',
  'percentile': 0.041552914493275805,
  'significant': True,
  'trait_id': 'facet_dutifulness'},
 {'category': 'personality',
  'name': 'Orderliness',
  'percentile': 0.8674746970454448,
  'significant': True,
  'trait_id': 'facet_orderliness'},
 {'category': 'personality',
  'name': 'Self-discipline',
  'percentile': 0.1329757632808503,
  'significant': True,
  'trait_id': 'facet_self_discipline'},
 {'category': 'personality',
  'name': 'Self-efficacy',
  'percentile': 0.16777654223353877,
  'significant': True,
  'trait_id': 'facet_self_efficacy'}]

**I don't necessarily need anything here.**

In [0]:
profile["personality"][2]["children"]

[{'category': 'personality',
  'name': 'Activity level',
  'percentile': 0.5296401678262467,
  'significant': True,
  'trait_id': 'facet_activity_level'},
 {'category': 'personality',
  'name': 'Assertiveness',
  'percentile': 0.0952716723151425,
  'significant': True,
  'trait_id': 'facet_assertiveness'},
 {'category': 'personality',
  'name': 'Cheerfulness',
  'percentile': 0.3870430243895838,
  'significant': True,
  'trait_id': 'facet_cheerfulness'},
 {'category': 'personality',
  'name': 'Excitement-seeking',
  'percentile': 0.7961856092352693,
  'significant': True,
  'trait_id': 'facet_excitement_seeking'},
 {'category': 'personality',
  'name': 'Outgoing',
  'percentile': 0.10137417440037588,
  'significant': True,
  'trait_id': 'facet_friendliness'},
 {'category': 'personality',
  'name': 'Gregariousness',
  'percentile': 0.05635805754081957,
  'significant': True,
  'trait_id': 'facet_gregariousness'}]

**I want assertiveness, cheerfulness, and outgoing.** At profile["personality"][2]["children"] 1,2,4.**

In [0]:
profile["personality"][3]["children"]

[{'category': 'personality',
  'name': 'Altruism',
  'percentile': 0.538223668884807,
  'significant': True,
  'trait_id': 'facet_altruism'},
 {'category': 'personality',
  'name': 'Cooperation',
  'percentile': 0.23109053962463755,
  'significant': True,
  'trait_id': 'facet_cooperation'},
 {'category': 'personality',
  'name': 'Modesty',
  'percentile': 0.4211426512168481,
  'significant': True,
  'trait_id': 'facet_modesty'},
 {'category': 'personality',
  'name': 'Uncompromising',
  'percentile': 0.5332614966874416,
  'significant': True,
  'trait_id': 'facet_morality'},
 {'category': 'personality',
  'name': 'Sympathy',
  'percentile': 0.9850128762364618,
  'significant': True,
  'trait_id': 'facet_sympathy'},
 {'category': 'personality',
  'name': 'Trust',
  'percentile': 0.33490127396513636,
  'significant': True,
  'trait_id': 'facet_trust'}]

**I want modesty and sympathy.** profile["personality"][3 ]["children"] 2,4

In [0]:
profile["personality"][4]["children"]

[{'category': 'personality',
  'name': 'Fiery',
  'percentile': 0.38635243247105927,
  'significant': True,
  'trait_id': 'facet_anger'},
 {'category': 'personality',
  'name': 'Prone to worry',
  'percentile': 0.42765965594186733,
  'significant': True,
  'trait_id': 'facet_anxiety'},
 {'category': 'personality',
  'name': 'Melancholy',
  'percentile': 0.9215424971788599,
  'significant': True,
  'trait_id': 'facet_depression'},
 {'category': 'personality',
  'name': 'Immoderation',
  'percentile': 0.5428585184312793,
  'significant': True,
  'trait_id': 'facet_immoderation'},
 {'category': 'personality',
  'name': 'Self-consciousness',
  'percentile': 0.8404856660550624,
  'significant': True,
  'trait_id': 'facet_self_consciousness'},
 {'category': 'personality',
  'name': 'Susceptible to stress',
  'percentile': 0.3295975826844284,
  'significant': True,
  'trait_id': 'facet_vulnerability'}]

**I want fiery and melancholy. profile["personality"][4]["children"] 0,2.**

**This gives me a total of 10 personality traits.**

In [0]:
# Testing putting those ten traits into a dictionary
traits = {}
traits["artistic"] = profile["personality"][0]["children"][1]["percentile"]
traits["emotion"] = profile["personality"][0]["children"][2]["percentile"]
traits["imagination"] = profile["personality"][0]["children"][3]["percentile"]

traits["assertive"] = profile["personality"][2]["children"][1]["percentile"]
traits["cheeful"] = profile["personality"][2]["children"][2]["percentile"]
traits["outgoing"] = profile["personality"][2]["children"][4]["percentile"]

traits["modesty"] = profile["personality"][3]["children"][2]["percentile"]
traits["sympathy"] = profile["personality"][3]["children"][4]["percentile"]

traits["fiery"] = profile["personality"][4]["children"][0]["percentile"]
traits["melancholy"] = profile["personality"][4]["children"][2]["percentile"]

traits

{'artistic': 0.9999827901171386,
 'assertive': 0.0952716723151425,
 'cheeful': 0.3870430243895838,
 'emotion': 0.9315941301830221,
 'fiery': 0.38635243247105927,
 'imagination': 0.9983841275049014,
 'melancholy': 0.9215424971788599,
 'modesty': 0.4211426512168481,
 'outgoing': 0.10137417440037588,
 'sympathy': 0.9850128762364618}

**The IBM Watson Personality Insights documentation states that 3000 words are needed for maximum precision, but that above 100 terms is acceptable for an analysis. Less than 100 terms will be an ApiException.**

In [0]:
# Grabbing the 10 chosen personality traits for each lyric
from ibm_watson import ApiException

personality = []

for lyric in df_songs_sentiment["lyrics"]:
  traits = {}

  try:
    profile = personality_insights.profile(
          lyric,
          accept='application/json',
          content_type='text/html',
    ).get_result()  

    traits["artistic"] = profile["personality"][0]["children"][1]["percentile"]
    traits["emotion"] = profile["personality"][0]["children"][2]["percentile"]
    traits["imagination"] = profile["personality"][0]["children"][3]["percentile"]

    traits["assertive"] = profile["personality"][2]["children"][1]["percentile"]
    traits["cheeful"] = profile["personality"][2]["children"][2]["percentile"]
    traits["outgoing"] = profile["personality"][2]["children"][4]["percentile"]

    traits["modesty"] = profile["personality"][3]["children"][2]["percentile"]
    traits["sympathy"] = profile["personality"][3]["children"][4]["percentile"]

    traits["fiery"] = profile["personality"][4]["children"][0]["percentile"]
    traits["melancholy"] = profile["personality"][4]["children"][2]["percentile"]

    personality.append(traits)

  # To continue the program if an ApiException occurrs. Will append an empty 
  # dictionary. The API requires at least 100 tokens to do an analysis, which
  # not all my lyrics might contain.
  except ApiException:
    personality.append({})

ERROR:root:The number of words 75 is less than the minimum number of words required for analysis: 100
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/ibm_cloud_sdk_core/base_service.py", line 229, in send
    response.status_code, error_message, http_response=response)
ibm_cloud_sdk_core.api_exception.ApiException: Error: The number of words 75 is less than the minimum number of words required for analysis: 100, Code: 400 , X-global-transaction-id: 6324869e339cf62ab63d5bb561412736
ERROR:root:The number of words 74 is less than the minimum number of words required for analysis: 100
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/ibm_cloud_sdk_core/base_service.py", line 229, in send
    response.status_code, error_message, http_response=response)
ibm_cloud_sdk_core.api_exception.ApiException: Error: The number of words 74 is less than the minimum number of words required for analysis: 100, Code: 400 , X-global-transacti

In [0]:
df_personality = json_normalize(personality)
df_personality.isnull().sum()

artistic       10
emotion        10
imagination    10
assertive      10
cheeful        10
outgoing       10
modesty        10
sympathy       10
fiery          10
melancholy     10
dtype: int64

**10 of my lyrics contained less than 100 words, but that is OK.**

In [0]:
df_personality

artistic   emotion  imagination  ...  sympathy     fiery  melancholy
0    0.999983  0.931594     0.998384  ...  0.985013  0.386352    0.921542
1    0.996511  0.986234     0.991554  ...  0.842244  0.157679    0.626636
2    0.906971  0.318395     0.994980  ...  0.785487  0.582447    0.669157
3    0.967339  0.229457     0.975867  ...  0.270318  0.787414    0.197281
4    0.980110  0.889895     0.977369  ...  0.875111  0.464351    0.871338
..        ...       ...          ...  ...       ...       ...         ...
826  0.930841  0.915972     0.965041  ...  0.768947  0.798722    0.982352
827  0.988442  0.748622     0.999587  ...  0.693261  0.935605    0.962687
828  0.995897  0.990308     0.999998  ...  0.966235  0.990844    0.999999
829  0.974639  0.804268     0.934779  ...  0.892613  0.147212    0.884749
830  0.982715  0.981525     0.897826  ...  0.796644  0.190062    0.950440

[831 rows x 10 columns]

In [0]:
df_songs_sentiment

artist  ... magnitude
0    Kendrick Lamar  ...       1.2
1    Kendrick Lamar  ...       1.3
2    Kendrick Lamar  ...       2.2
3    Kendrick Lamar  ...       1.3
4    Kendrick Lamar  ...       1.1
..              ...  ...       ...
826          Halsey  ...       1.4
827          Halsey  ...       1.0
828          Halsey  ...       1.7
829          Halsey  ...       1.1
830          Halsey  ...       0.6

[831 rows x 9 columns]

In [0]:
df_personality_sentiment = df_songs_sentiment.merge(df_personality, 
                          right_index=True, left_index=True, how="outer")
df_personality_sentiment

artist  ... melancholy
0    Kendrick Lamar  ...   0.921542
1    Kendrick Lamar  ...   0.626636
2    Kendrick Lamar  ...   0.669157
3    Kendrick Lamar  ...   0.197281
4    Kendrick Lamar  ...   0.871338
..              ...  ...        ...
826          Halsey  ...   0.982352
827          Halsey  ...   0.962687
828          Halsey  ...   0.999999
829          Halsey  ...   0.884749
830          Halsey  ...   0.950440

[831 rows x 19 columns]

In [0]:
# I am not interested in songs that I don't have personality analysis for. Also
# songs shorter than 100 words aren't that great for predictiveness or for
# random song generation.
df_personality_sentiment = df_personality_sentiment \
            [df_personality_sentiment["artistic"].notnull()]
df_personality_sentiment = df_personality_sentiment.reset_index().iloc[:,1:]
df_personality_sentiment

artist  ... melancholy
0    Kendrick Lamar  ...   0.921542
1    Kendrick Lamar  ...   0.626636
2    Kendrick Lamar  ...   0.669157
3    Kendrick Lamar  ...   0.197281
4    Kendrick Lamar  ...   0.871338
..              ...  ...        ...
816          Halsey  ...   0.982352
817          Halsey  ...   0.962687
818          Halsey  ...   0.999999
819          Halsey  ...   0.884749
820          Halsey  ...   0.950440

[821 rows x 19 columns]

# Final Cleaning

In [0]:
df_personality_sentiment.head()

artist  ... melancholy
0  Kendrick Lamar  ...   0.921542
1  Kendrick Lamar  ...   0.626636
2  Kendrick Lamar  ...   0.669157
3  Kendrick Lamar  ...   0.197281
4  Kendrick Lamar  ...   0.871338

[5 rows x 19 columns]

**The artists and track columns are unnecessary, and I would like to transform duration to seconds instead of milliseconds.**

**Also at this point, I want the boolean explicit column to be 0 and 1.**

In [0]:
df_personality_sentiment["explicit"] = \
            df_personality_sentiment["explicit"].astype("int64")
df_personality_sentiment.head()

uri          artist  ... melancholy duration_s
0  spotify:track:5qN4HFkapdAOV94XPryVof  Kendrick Lamar  ...   0.921542    130.613
1  spotify:track:3GCdLUSnKSMJhs4Tj6CV3s  Kendrick Lamar  ...   0.626636    232.186
2  spotify:track:0DJBgBiYeSn6n1AXAkFVE8  Kendrick Lamar  ...   0.669157    208.760
3  spotify:track:5cXg9IQS34FzLVdHhp7hu7  Kendrick Lamar  ...   0.197281    221.826
4  spotify:track:77UjLW8j5UAGAGVGhR5oUK  Kendrick Lamar  ...   0.871338    211.440

[5 rows x 31 columns]

**Above is my final sentiment personality dataset. Now I look at the first dataset without lyrics. I want to do the same things, remove artists and track_number, transform duration to s, and map explicit to 0 and 1.**

In [0]:
df_songs.head()

artist  ...   genre
uri                                                   ...        
spotify:track:5qN4HFkapdAOV94XPryVof  Kendrick Lamar  ...  hiphop
spotify:track:3GCdLUSnKSMJhs4Tj6CV3s  Kendrick Lamar  ...  hiphop
spotify:track:4LmAnpjlhWTahvRkYR8xJa  Kendrick Lamar  ...  hiphop
spotify:track:5jyyPsIGM2yqkZN9R3TmvN  Kendrick Lamar  ...  hiphop
spotify:track:7bUcBztfGqO7cSI2gMZeCI  Kendrick Lamar  ...  hiphop

[5 rows x 21 columns]

In [0]:
df_songs["explicit"] = df_songs["explicit"].astype("int64")
df_songs["duration_s"] = df_songs["duration_ms"] / 1000
df_songs = df_songs.drop(
    columns = ["artists", "track_number","duration_ms"], axis=1
)
df_songs.head()

artist  ... duration_s
uri                                                   ...           
spotify:track:5qN4HFkapdAOV94XPryVof  Kendrick Lamar  ...    130.613
spotify:track:3GCdLUSnKSMJhs4Tj6CV3s  Kendrick Lamar  ...    232.186
spotify:track:4LmAnpjlhWTahvRkYR8xJa  Kendrick Lamar  ...    267.426
spotify:track:5jyyPsIGM2yqkZN9R3TmvN  Kendrick Lamar  ...    238.893
spotify:track:7bUcBztfGqO7cSI2gMZeCI  Kendrick Lamar  ...    180.893

[5 rows x 19 columns]

In [9]:
df_personality_sentiment.head()

artist  ... melancholy
0  Kendrick Lamar  ...   0.921542
1  Kendrick Lamar  ...   0.626636
2  Kendrick Lamar  ...   0.669157
3  Kendrick Lamar  ...   0.197281
4  Kendrick Lamar  ...   0.871338

[5 rows x 19 columns]

**I will be trying to predict popularity, so I want popularity in the personality sentiment dataframe.**

In [0]:
df_merge = df_songs[["artist", "name", "album", "popularity"]]
df_personality_sentiment = df_personality_sentiment.merge(
    df_merge, on=["artist", "name", "album"], how="left")

In [11]:
df_personality_sentiment.isnull().sum()

artist          0
album           0
name            0
explicit        0
genre           0
chain_lyrics    0
lyrics          0
polarity        0
magnitude       0
artistic        0
emotion         0
imagination     0
assertive       0
cheeful         0
outgoing        0
modesty         0
sympathy        0
fiery           0
melancholy      0
popularity      0
dtype: int64

**Now I can save these two dataframes.**


In [0]:
df_songs.to_csv("/content/drive/My Drive/df_songs.csv", index=False)

In [0]:
df_personality_sentiment.to_csv(
    "/content/drive/My Drive/df_personality_sentiment.csv", index=False)